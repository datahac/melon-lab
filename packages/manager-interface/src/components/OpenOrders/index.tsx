import React from 'react';
import Composer from 'react-composer';
import { withPropsOnChange } from 'recompose';
import OpenOrders from '~/components/OpenOrders';
import { OpenOrdersQuery, OpenOrdersMutation } from './data/openOrders';
import * as R from 'ramda';

// TODO: This mapping shouldn't be necessary?
const withMappedOrders = withPropsOnChange(['orders'], props => ({
  orders: props.orders.map(order => ({
    ...order,
    buyHowMuch: order.buy.howMuch,
    buySymbol: order.buy.symbol,
    sellHowMuch: order.sell.howMuch,
    sellSymbol: order.sell.symbol,
  })),
}));

const OpenOrdersMapped = withMappedOrders(OpenOrders);

export default class OpenOrdersContainer extends React.PureComponent {
  render() {
    return (
      <Composer components={[
        <OpenOrdersQuery address={this.props.address} />,
        ({ render }) => (
          <OpenOrdersMutation>{(a, b) => render([a, b])}</OpenOrdersMutation>
        ),
      ]}>
        {([openOrdersProps, [cancelOrder]]) => {
          const orders = R.pathOr([], ['data', 'openOrders'])(openOrdersProps);

          return (
            <OpenOrdersMapped
              {...this.props}
              orders={orders}
              loading={openOrdersProps.loading}
              onClick={(orderId, makerAssetSymbol, takerAssetSymbol) => {
                cancelOrder({
                  variables: {
                    orderId,
                    fundAddress: this.props.address,
                    makerAssetSymbol,
                    takerAssetSymbol,
                  },
                });
              }}
            />
          );
        }}
      </Composer>
    )
  }
}
export default OpenOrdersContainer;
