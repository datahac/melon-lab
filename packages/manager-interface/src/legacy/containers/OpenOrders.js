import moment from 'moment';
import { connect } from 'react-redux';
import { lifecycle } from 'recompose';
import { actions } from '../actions/openOrders';
import OpenOrders from '@melonproject/manager-components/components/OpenOrders';
import displayNumber from '../utils/displayNumber';
import isSameAddress from '../utils/isSameAddress';

const mapStateToProps = state => ({
  ...state.openOrders,
  orders: state.openOrders.orders.map(order => ({
    buyHowMuch: displayNumber(order.buyHowMuch),
    buySymbol: order.buySymbol,
    id: order.exchangeOrderId,
    price: displayNumber(order.price),
    sellHowMuch: displayNumber(order.sellHowMuch),
    sellSymbol: order.sellSymbol,
    timestamp: moment(order.timestamp).format('D. MMM YYYY HH:mm'),
    type: order.type,
  })),
  isReadyToTrade: state.app.isReadyToTrade,
  isManager:
    state.app.isReadyToInteract &&
    isSameAddress(state.ethereum.account, state.fund.owner),
});

const mapDispatchToProps = dispatch => ({
  getOpenOrders: () => {
    dispatch(actions.getOpenOrders());
  },
  onClick: (orderId, makerAssetSymbol, takerAssetSymbol) => {
    dispatch(actions.cancelOrder(orderId, makerAssetSymbol, takerAssetSymbol));
  },
});

const OpenOrdersWithLifecycle = lifecycle({
  componentDidMount() {
    this.props.getOpenOrders();
  },
})(OpenOrders);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(OpenOrdersWithLifecycle);
