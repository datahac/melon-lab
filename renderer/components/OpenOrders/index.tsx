import React from 'react';
import { withPropsOnChange } from 'recompose';
import OpenOrders from '~/components/OpenOrders';
import { OpenOrdersQuery } from './data/openOrders';
import * as R from 'ramda';
import CancelOrder from '../CancelOrder';

// TODO: This mapping shouldn't be necessary?
const withMappedOrders = withPropsOnChange(['orders'], props => {
  return {
    orders: props.orders.map(order => {
      const buy = order.type === 'ASK' ? order.trade.quote : order.trade.base;
      const sell = order.type === 'ASK' ? order.trade.base : order.trade.quote;

      return {
        ...order,
        timestamp: order.metadata.timestamp,
        buyHowMuch: buy.quantity,
        buySymbol: buy.token.symbol,
        sellHowMuch: sell.quantity,
        sellSymbol: sell.token.symbol,
      };
    }),
  };
});

const OpenOrdersMapped = withMappedOrders(OpenOrders);

export default class OpenOrdersContainer extends React.Component {
  state = { selectedOrder: null };

  render() {
    return (
      <OpenOrdersQuery fundAddress={this.props.address}>
        {props => {
          const orders = R.pathOr([], ['data', 'openOrders'])(props);

          return (
            <React.Fragment>
              <OpenOrdersMapped
                {...this.props}
                orders={orders}
                loading={props.loading}
                onClick={orderId => {
                  this.setState({
                    selectedOrder: {
                      id: orderId,
                    },
                  });
                }}
              />
              <CancelOrder
                values={this.state.selectedOrder}
                refresh={() => props.refetch(props.variable)}
                unset={() => this.setState({ selectedOrder: null })}
              />
            </React.Fragment>
          );
        }}
      </OpenOrdersQuery>
    );
  }
}
