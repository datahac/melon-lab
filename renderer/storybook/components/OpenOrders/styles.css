.open-orders {
  border-top: 1px solid var(--color-border);
}

.open-orders__table-wrap {
  overflow-x: auto;
}

.open-orders__cell--red {
  color: var(--color-loss);
}

.open-orders__cell--green {
  color: var(--color-profit);
}

.open-orders__no-items {
  margin: var(--space-xs) var(--space-s);
}

.open-orders__type--bid {
  color: var(--color-profit);
}

.open-orders__type--ask {
  color: var(--color-loss);
}
