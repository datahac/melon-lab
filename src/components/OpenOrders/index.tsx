import React, { useState } from 'react';
import OpenOrders from '~/components/OpenOrders';
import { OpenOrdersQuery } from './data/openOrders';
import * as R from 'ramda';
import CancelOrder from '+/components/CancelOrder';

const OpenOrdersContainer = ({ address, canInteract, isManager }) => {
  const [selectedOrder, setSelectedOrder] = useState(null);

  return (
    <OpenOrdersQuery fundAddress={address}>
      {queryProps => (
        <React.Fragment>
          <OpenOrders
            canInteract={canInteract}
            isManager={isManager}
            loading={queryProps.loading}
            orders={R.pathOr([], ['data', 'openOrders'], queryProps)}
            onClick={setSelectedOrder}
          />
          <CancelOrder values={selectedOrder} setSelectedOrder={setSelectedOrder} />
        </React.Fragment>
      )}
    </OpenOrdersQuery>
  );
};

export default OpenOrdersContainer;
