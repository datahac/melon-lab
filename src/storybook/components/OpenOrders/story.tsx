import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import React from 'react';
import OpenOrders from './index';

const data = {
  isManager: true,
  isReadyToTrade: true,
  orders: [
    {
      buyHowMuch: '2.1180',
      buySymbol: 'MLN',
      id: 12322,
      price: '0.0672',
      sellHowMuch: '0.1423',
      sellSymbol: 'WETH',
      timestamp: '25. Jul 2018 11:11',
      type: 'sell',
    },
    {
      buyHowMuch: '2.1180',
      buySymbol: 'MLN',
      id: 29468,
      price: '0.0672',
      sellHowMuch: '0.1423',
      sellSymbol: 'WETH',
      timestamp: '25. Jul 2018 11:11',
      type: 'sell',
    },
  ],
  onClick: action('onClick'),
};

storiesOf('Components|Open Orders', module).add('Default', () => {
  return <OpenOrders {...data} />;
});
