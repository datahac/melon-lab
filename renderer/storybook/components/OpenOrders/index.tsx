import React, { StatelessComponent } from 'react';
import Button from '~/blocks/Button';
import {
  CellBody,
  CellHead,
  Row,
  Table,
  TableBody,
  TableHead,
} from '~/blocks/Table';
import displayQuantity from '~/shared/utils/displayQuantity';

import styles from './styles.css';

export interface Order {
  buySymbol: string;
  id: number;
  price: string;
  quantity: string;
  timestamp: string;
  type: string;
}

export interface OpenOrdersProps {
  isManager?: boolean;
  canInteract?: boolean;
  onClick: (id) => void;
  orders?: Order[];
}

export const OpenOrders: StatelessComponent<OpenOrdersProps> = ({
  isManager,
  canInteract,
  onClick,
  orders = [],
}) => {
  return (
    <div className="open-orders">
      <style jsx={true}>{styles}</style>
      <div className="open-orders__table-wrap">
        {orders.length > 0 ? (
          <Table>
            <TableHead>
              <Row isHead={true} size={isManager && 'small'}>
                <CellHead noPadding={false}>Buy</CellHead>
                <CellHead>Sell</CellHead>
                <CellHead>Price</CellHead>
                <CellHead noPadding={false}>Quantity</CellHead>
                {isManager && <CellHead noPadding={false} />}
              </Row>
            </TableHead>
            <TableBody>
              {orders &&
                orders.map(order => {
                  const buy =
                    order.type === 'ASK' ? order.trade.quote : order.trade.base;
                  const sell =
                    order.type === 'ASK' ? order.trade.base : order.trade.quote;

                  return (
                    <Row key={order.id} size="medium">
                      <CellBody noPadding={false}>{sell.token.symbol}</CellBody>
                      <CellBody>{buy.token.symbol}</CellBody>
                      <CellBody>
                        {order.price && displayQuantity(order.price)}
                      </CellBody>
                      <CellBody>
                        {order.volume && displayQuantity(order.volume)}
                      </CellBody>
                      {isManager && (
                        <CellBody noPadding={false} width={50}>
                          <Button
                            icon="cross"
                            size="small"
                            style="clear"
                            onClick={() => onClick(order)}
                            disabled={!canInteract}
                          />
                        </CellBody>
                      )}
                    </Row>
                  );
                })}
            </TableBody>
          </Table>
        ) : (
          <p className="open-orders__no-items">No open orders</p>
        )}
      </div>
    </div>
  );
};

export default OpenOrders;
