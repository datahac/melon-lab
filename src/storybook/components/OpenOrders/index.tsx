import classNames from 'classnames';
import React, { StatelessComponent } from 'react';
import Button from '~/blocks/Button';
import {
  CellBody,
  CellHead,
  Row,
  Table,
  TableBody,
  TableHead,
} from '~/blocks/Table';
import format from 'date-fns/format';
import displayNumber from '~/utils/displayNumber';

import styles from './styles.css';

export interface Order {
  buyHowMuch: string;
  buySymbol: string;
  id: number;
  price: string;
  sellHowMuch: string;
  sellSymbol: string;
  timestamp: string;
  type: string;
}

export interface OpenOrdersProps {
  isManager?: boolean;
  isReadyToTrade?: boolean;
  onClick: (id, buySymbol, sellSymbol) => void;
  orders?: Order[];
}

export const OpenOrders: StatelessComponent<OpenOrdersProps> = ({
  isManager,
  isReadyToTrade,
  onClick,
  orders = [],
}) => {
  const typeCellClassNames = (type: string) =>
    classNames(
      'open-orders__cell',
      {
        'open-orders__cell--red': type === 'sell',
      },
      {
        'open-orders__cell--green': type === 'buy',
      },
    );

  return (
    <div className="open-orders">
      <style jsx>{styles}</style>
      <div className="open-orders__table-wrap">
        {orders.length > 0 ? (
          <Table>
            <TableHead>
              <Row isHead={true} size={isManager && 'small'}>
                <CellHead noPadding={false}>Time</CellHead>
                <CellHead>Id</CellHead>
                <CellHead>Type</CellHead>
                <CellHead>Buy</CellHead>
                <CellHead>Sell</CellHead>
                <CellHead>Price</CellHead>
                <CellHead>Buy Quantity</CellHead>
                <CellHead noPadding={false}>Sell Quantity</CellHead>
                {isManager && <CellHead noPadding={false} />}
              </Row>
            </TableHead>
            <TableBody>
              {orders &&
                orders.map(order => {
                  return (
                    <Row key={order.id} size={isManager && 'small'}>
                      <CellBody noPadding={false}>
                        {format(
                          parseInt(order.timestamp),
                          'DD. MMM YYYY HH:mm',
                        )}
                      </CellBody>
                      <CellBody>{order.id}</CellBody>
                      <CellBody>
                        <span className={typeCellClassNames(order.type)}>
                          {order.type}
                        </span>
                      </CellBody>
                      <CellBody>{order.sellSymbol}</CellBody>
                      <CellBody>{order.buySymbol}</CellBody>
                      <CellBody>{displayNumber(order.price)}</CellBody>
                      <CellBody>{displayNumber(order.sellHowMuch)}</CellBody>
                      <CellBody noPadding={false}>
                        {displayNumber(order.buyHowMuch)}
                      </CellBody>
                      {isManager && (
                        <CellBody noPadding={false}>
                          <Button
                            size="small"
                            style="secondary"
                            onClick={() =>
                              onClick(
                                order.id,
                                order.buySymbol,
                                order.sellSymbol,
                              )
                            }
                            disabled={!isReadyToTrade}
                          >
                            Cancel
                          </Button>
                        </CellBody>
                      )}
                    </Row>
                  );
                })}
            </TableBody>
          </Table>
        ) : (
          <p className="open-orders__no-items">No open orders</p>
        )}
      </div>
    </div>
  );
};

export default OpenOrders;
