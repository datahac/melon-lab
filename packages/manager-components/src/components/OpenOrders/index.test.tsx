import React from 'react';
import OpenOrders from './index';

const data = {
  isManager: true,
  isReadyToTrade: true,
  orders: [
    {
      buyHowMuch: '2.1180',
      buySymbol: 'MLN-T',
      id: 12322,
      price: '0.0672',
      sellHowMuch: '0.1423',
      sellSymbol: 'WETH-T',
      timestamp: '25. Jul 2018 11:11',
      type: 'sell',
    },
    {
      buyHowMuch: '2.1180',
      buySymbol: 'MLN-T',
      id: 29468,
      price: '0.0672',
      sellHowMuch: '0.1423',
      sellSymbol: 'WETH-T',
      timestamp: '25. Jul 2018 11:11',
      type: 'sell',
    },
  ],
  onClick: () => null,
};

describe('OpenOrders', () => {
  const defaultElement = <OpenOrders {...data} />;

  it('should render correctly', () => {
    const wrapper = shallow(defaultElement);
    expect(wrapper).toMatchSnapshot();
  });
});
