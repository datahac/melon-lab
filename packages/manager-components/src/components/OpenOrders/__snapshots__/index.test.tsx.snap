// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenOrders should render correctly 1`] = `
<div
  className="jsx-__scopedHash open-orders"
>
  <JSXStyle
    css="__scoped"
    styleId="__scopedHash"
  />
  <h3
    className="jsx-__scopedHash"
  >
    Open orders
  </h3>
  <div
    className="jsx-__scopedHash open-orders__table-wrap"
  >
    <Table>
      <TableHead>
        <Row
          isHead={true}
          size="small"
        >
          <CellHead>
            Time
          </CellHead>
          <CellHead>
            Id
          </CellHead>
          <CellHead>
            Type
          </CellHead>
          <CellHead>
            Buy
          </CellHead>
          <CellHead>
            Sell
          </CellHead>
          <CellHead>
            Price
          </CellHead>
          <CellHead>
            Buy Quantity
          </CellHead>
          <CellHead>
            Sell Quantity
          </CellHead>
          <CellHead />
        </Row>
      </TableHead>
      <TableBody>
        <Row
          key="12322"
          size="small"
        >
          <CellBody>
            25. Jul 2018 11:11
          </CellBody>
          <CellBody>
            12322
          </CellBody>
          <CellBody>
            <span
              className="jsx-__scopedHash open-orders__cell open-orders__cell--red"
            >
              sell
            </span>
          </CellBody>
          <CellBody>
            WETH-T
          </CellBody>
          <CellBody>
            MLN-T
          </CellBody>
          <CellBody>
            0.0672
          </CellBody>
          <CellBody>
            0.1423
          </CellBody>
          <CellBody>
            2.1180
          </CellBody>
          <CellBody>
            <Button
              buttonValue={
                Object {
                  "buySymbol": "MLN-T",
                  "id": 12322,
                  "sellSymbol": "WETH-T",
                }
              }
              disabled={false}
              onClick={[Function]}
              size="small"
              style="secondary"
            >
              Cancel
            </Button>
          </CellBody>
        </Row>
        <Row
          key="29468"
          size="small"
        >
          <CellBody>
            25. Jul 2018 11:11
          </CellBody>
          <CellBody>
            29468
          </CellBody>
          <CellBody>
            <span
              className="jsx-__scopedHash open-orders__cell open-orders__cell--red"
            >
              sell
            </span>
          </CellBody>
          <CellBody>
            WETH-T
          </CellBody>
          <CellBody>
            MLN-T
          </CellBody>
          <CellBody>
            0.0672
          </CellBody>
          <CellBody>
            0.1423
          </CellBody>
          <CellBody>
            2.1180
          </CellBody>
          <CellBody>
            <Button
              buttonValue={
                Object {
                  "buySymbol": "MLN-T",
                  "id": 29468,
                  "sellSymbol": "WETH-T",
                }
              }
              disabled={false}
              onClick={[Function]}
              size="small"
              style="secondary"
            >
              Cancel
            </Button>
          </CellBody>
        </Row>
      </TableBody>
    </Table>
  </div>
</div>
`;
